gson/src/main/java/com/google/gson/Gson.java

```diff
@@ -523,46 +524,76 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
       return adapter;
     }
 
-    Map<TypeToken<?>, FutureTypeAdapter<?>> threadCalls = calls.get();
-    boolean requiresThreadLocalCleanup = false;
+    LinkedHashMap<TypeToken<?>, TypeAdapter<?>> threadCalls = calls.get();
+    boolean isInitialAdapterRequest = false;
     if (threadCalls == null) {
-      threadCalls = new HashMap<>();
+      threadCalls = new LinkedHashMap<>();
       calls.set(threadCalls);
-      requiresThreadLocalCleanup = true;
+      isInitialAdapterRequest = true;
     }
 
     // the key and value type parameters always agree
     @SuppressWarnings("unchecked")
-    FutureTypeAdapter<T> ongoingCall = (FutureTypeAdapter<T>) threadCalls.get(type);
+    TypeAdapter<T> ongoingCall = (TypeAdapter<T>) threadCalls.get(type);
     if (ongoingCall != null) {
       return ongoingCall;
     }
 
+    int existingAdaptersCount = threadCalls.size();
+    boolean foundCandidate = false;
     try {
       FutureTypeAdapter<T> call = new FutureTypeAdapter<>();
       threadCalls.put(type, call);
 
       for (TypeAdapterFactory factory : factories) {
         TypeAdapter<T> candidate = factory.create(this, type);
         if (candidate != null) {
-          @SuppressWarnings("unchecked")
-          TypeAdapter<T> existingAdapter = (TypeAdapter<T>) typeTokenCache.putIfAbsent(type, candidate);
-          // If other thread concurrently added adapter prefer that one instead
-          if (existingAdapter != null) {
-            candidate = existingAdapter;
-          }
-
           call.setDelegate(candidate);
+          // Replace future adapter with actual adapter
+          threadCalls.put(type, candidate);
+
+          if (isInitialAdapterRequest) {
+            // Publish resolved adapters to all threads
+            // Can only do this for the initial request because cyclic dependency TypeA -> TypeB -> TypeA
+            // would otherwise publish adapter for TypeB which uses not yet resolved adapter for TypeA
+            // See https://github.com/google/gson/issues/625
+            for (Map.Entry<TypeToken<?>, TypeAdapter<?>> resolvedAdapterEntry : threadCalls.entrySet()) {
+              typeTokenCache.putIfAbsent(resolvedAdapterEntry.getKey(), resolvedAdapterEntry.getValue());
+            }
+
+            @SuppressWarnings("unchecked")
+            TypeAdapter<T> actualAdapter = (TypeAdapter<T>) typeTokenCache.get(type);
```

Comment:
This separate `get` call is needed to preserve the behavior of #2153. It retrieves the adapter `putIfAbsent` added above (since it is also called with `candidate` as value), or the existing one in case another thread concurrently put the adapter into the map.

Note that this cannot guarantee that for nested adapters the same adapter instance is used as well, e.g.:
```
A
  X
B
  X
```
Here when requesting adapters for `A` and `B` concurrently they might refer to different instances of the adapter for `X`. Though that is probably acceptable.

Reply:
That is fine for me. When I created #2153 (which introduced this `putIfAbsent` functionality) I thought it might be nice to have, but as far as I know there was no issue explicitly requesting this.